drivers/display/display_sh1122.c

```diff
@@ -0,0 +1,538 @@
+	}
+	err = config->write_cmd(dev, SH1122_SET_VSEGM, &config->precharge_voltage, 1);
+	if (err < 0) {
+		return err;
+	}
+	err = config->write_cmd(dev, SH1122_SET_VCOMH, &config->vcomh_voltage, 1);
+	if (err < 0) {
+		return err;
+	}
+	return config->write_cmd(dev, SH1122_SET_DCDC, &config->dc_dc, 1);
+}
+
+static int sh1122_resume(const struct device *dev)
+{
+	const struct sh1122_config *config = dev->config;
+	int err;
+
+	err = config->write_cmd(dev, SH1122_DISPLAY_ON, NULL, 0);
+	if (err < 0) {
+		return err;
+	}
+	return config->release_bus(dev);
+}
+
+static int sh1122_suspend(const struct device *dev)
+{
+	const struct sh1122_config *config = dev->config;
+	int err;
+
+	err = config->write_cmd(dev, SH1122_DISPLAY_OFF, NULL, 0);
+	if (err < 0) {
+		return err;
+	}
+	return config->release_bus(dev);
+}
+
+/* Convert what the conversion buffer can hold to pixelx (3:0) and pixelx+1 (7:4) */
+
+#if DT_HAS_COMPAT_ON_BUS_STATUS_OKAY(sinowealth_sh1122, mipi_dbi)
+static int sh1122_write_pixels_mipi(const struct device *dev, const uint8_t *buf,
+				    const struct display_buffer_descriptor *desc)
+{
+	const struct sh1122_config *config = dev->config;
+	struct display_buffer_descriptor mipi_desc;
+
+	mipi_desc.buf_size = desc->width / 2;
+	mipi_desc.pitch = desc->pitch / 2;
+	mipi_desc.width = desc->width;
+	mipi_desc.height = 1;
+
+	/* This is the wrong format, but it doesn't matter to almost all mipi drivers */
+	return mipi_dbi_write_display(config->mipi_dev, &config->dbi_config, config->conversion_buf,
+				      &mipi_desc, PIXEL_FORMAT_L_8);
+}
+
+static int sh1122_release_bus_mipi(const struct device *dev)
+{
+	const struct sh1122_config *config = dev->config;
+
+	return mipi_dbi_release(config->mipi_dev, &config->dbi_config);
+}
+#endif
+
+#if DT_HAS_COMPAT_ON_BUS_STATUS_OKAY(sinowealth_sh1122, i2c)
+static int sh1122_write_pixels_i2c(const struct device *dev, const uint8_t *buf,
+				   const struct display_buffer_descriptor *desc)
+{
+	const struct sh1122_config *config = dev->config;
+	int ret, i;
+	int total = 0;
+
+	return i2c_burst_write_dt(&config->i2c, SH1122_CONTROL_ALL_BYTES_DATA,
+				  config->conversion_buf, desc->width / 2);
+}
+
+static int sh1122_release_bus_i2c(const struct device *dev)
+{
+	return 0;
+}
+#endif
+
+static int sh1122_write(const struct device *dev, const uint16_t x, const uint16_t y,
+			const struct display_buffer_descriptor *desc, const void *buf)
+{
+	const struct sh1122_config *config = dev->config;
+	int err;
+	size_t buf_len;
+	int pixel_count = desc->width * desc->height;
+	int total = 0;
+	uint8_t ybuf = y;
+
+	if (desc->pitch < desc->width) {
+		LOG_ERR("Pitch is smaller than width");
+		return -1;
+	}
+
+	/* Following the datasheet, in the GDDRAM, two segment are split in one register */
+	buf_len = MIN(desc->buf_size, desc->height * desc->width / 2);
+	if (buf == NULL || buf_len == 0U) {
+		LOG_ERR("Display buffer is not available");
+		return -1;
+	}
+
+	if (desc->pitch > desc->width) {
```

Comment:
Together with the above check this means you could simplify to `desc->pitch != desc->width`

Reply:
Done